File: GGAndroid_2.0/GGDemo/src/main/java/io/ganguo/demo/viewmodel/page/TabPageViewModel.java
package io.ganguo.demo.viewmodel.page;

import android.view.View;
import android.view.ViewGroup;

import io.ganguo.demo.R;
import io.ganguo.demo.viewmodel.item.ItemDemoVModel;
import io.ganguo.demo.viewmodel.item.ItemFooterVModel;
import io.ganguo.utils.common.ToastHelper;
import io.ganguo.vmodel.BaseViewModel;
import io.ganguo.vmodel.ViewModelHelper;
import io.ganguo.library.ui.view.ActivityInterface;
import io.ganguo.utils.util.Randoms;
import io.ganguo.viewmodel.common.HFSRecyclerViewModel;
import io.ganguo.viewmodel.common.SimpleViewModel;
import io.ganguo.viewmodel.databinding.IncludeHfSwipeRecyclerBinding;
import io.reactivex.functions.Consumer;

/**
 * Created by leo on 16/7/11.
 * TabLayout - 对应ViewModel
 */
public class TabPageViewModel extends HFSRecyclerViewModel<ActivityInterface<IncludeHfSwipeRecyclerBinding>> {

    @Override
    public void initHeader(ViewGroup container) {
        super.initHeader(container);
    }

    @Override
    public void initFooter(ViewGroup container) {
        super.initFooter(container);
        ItemFooterVModel testFooterViewModel = new ItemFooterVModel(onClear());
        ViewModelHelper.bind(container, this, testFooterViewModel);
    }

    @Override
    public void initEmpty(ViewGroup container) {
        super.initEmpty(container);
        SimpleViewModel emptyViewModel = new SimpleViewModel(R.layout.include_test_empty);
        ViewModelHelper.bind(container, this, emptyViewModel);
    }

    @Override
    public void onViewAttached(View view) {
        getRecyclerView().setFocusableInTouchMode(false);
        onRefresh();
    }

    @Override
    public void onRefresh() {
        super.onRefresh();
        getAdapter().disableLoadMore();
        loadDelay(true);
    }

    @Override
    public void onLoadMore() {
        super.onLoadMore();
        loadDelay(false);
    }

    /**
     * Clear data
     */
    public Consumer<View> onClear() {
        return view -> {
            getAdapter().clear();
            getAdapter().disableLoadMore();
            getAdapter().notifyDiffUtilSetDataChanged();
            toggleEmptyView();
        };
    }

    private void loadDelay(final boolean isRefresh) {
        if (isRefresh) {
            getAdapter().clear();
        }
        for (Integer i = 0; i < 20; i++) {
            getAdapter().add(createItemDemoVModel());
        }
        getAdapter().notifyDiffUtilSetDataChanged();
        toggleEmptyView();
    }

    /**
     * function: create Item ViewModel
     *
     * @return
     */
    public BaseViewModel createItemDemoVModel() {
        String content = Randoms.getRandomCapitalLetters(12);
        return new ItemDemoVModel<String>()
                .setBtnText("click")
                .setData(content)
                .setContent(content)
                .setClickAction(s -> ToastHelper.showMessage("click"));
    }

    @Override
    public void onAdapterNotifyComplete() {
        super.onAdapterNotifyComplete();
        getAdapter().enableLoadMore();
    }
}

File: GGAndroid_3.0/main/app-mvvm-v1-sample/src/main/java/io/ganguo/demo/viewmodel/page/TabPageViewModel.kt
package io.ganguo.demo.viewmodel.page

import android.view.View
import android.view.ViewGroup

import com.scwang.smart.refresh.layout.api.RefreshLayout

import io.ganguo.demo.viewmodel.item.ItemDemoVModel
import io.ganguo.demo.viewmodel.item.ItemFooterVModel
import io.ganguo.viewmodel.core.viewinterface.ActivityInterface
import io.ganguo.utils.helper.ToastHelper
import io.ganguo.utils.util.Randoms
import io.ganguo.viewmodel.pack.common.HFSRecyclerViewModel
import io.ganguo.viewmodel.databinding.IncludeHfSwipeRecyclerBinding
import io.ganguo.viewmodel.core.BaseViewModel
import io.ganguo.viewmodel.core.ViewModelHelper
import io.reactivex.functions.Consumer

/**
 * Created by leo on 16/7/11.
 * TabLayout - 对应ViewModel
 */
class TabPageViewModel : HFSRecyclerViewModel<ActivityInterface<IncludeHfSwipeRecyclerBinding>>() {

    override fun initFooter(container: ViewGroup) {
        super.initFooter(container)
        val testFooterViewModel = ItemFooterVModel(onClear())
        ViewModelHelper.bind(container, this, testFooterViewModel)
    }


    override fun onViewAttached(view: View) {
        recyclerView.isFocusableInTouchMode = false
        loadDelay(true)
    }


    override fun onLoadMore(refreshLayout: RefreshLayout) {
        super.onLoadMore(refreshLayout)
        loadDelay(false)
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        super.onRefresh(refreshLayout)
        loadDelay(true)
    }

    /**
     * Clear data
     */
    private fun onClear(): Consumer<View> {
        return Consumer {
            adapter.clear()
            adapter.notifySetDataDiffChanged()
            toggleEmptyView()
        }
    }

    private fun loadDelay(isRefresh: Boolean) {
        if (isRefresh) {
            adapter.clear()
        }
        for (i in 0..19) {
            adapter.add(newItemDemoVModel())
        }
        adapter.notifySetDataDiffChanged()
        toggleEmptyView()
    }

    /**
     * 生成 Item ViewModel
     *
     * @return
     */
    private fun newItemDemoVModel(): BaseViewModel<*> {
        val content = Randoms.getRandomCapitalLetters(12)
        return ItemDemoVModel<String>()
                .setBtnText("click")
                .setDataObj(content)
                .setContent(content)
                .setClickAction {
                    ToastHelper.showMessage("click")
                }
    }

}
